apis/v1beta1/config.go

```diff
@@ -445,67 +447,133 @@ const (
+		if isPortEnvVar {
+			errMsg := fmt.Sprintf("couldn't determine metrics port from configuration: %s",
+				telemetry.Metrics.Address)
+			logger.Info(errMsg)
+			return "", 0, errors.New(errMsg)
+		}
 
-	// The regex below matches on strings that end with a colon followed by 1 or more numbers (representing the port).
-	const explicitPortRegex = `:(\d+$)`
-	explicitPortMatches := regexp.MustCompile(explicitPortRegex).FindStringSubmatch(telemetry.Metrics.Address)
-	if len(explicitPortMatches) <= 1 {
-		return telemetry.Metrics.Address, defaultServicePort, nil
-	}
+		// The regex below matches on strings that end with a colon followed by 1 or more numbers (representing the port).
+		const explicitPortRegex = `:(\d+$)`
+		explicitPortMatches := regexp.MustCompile(explicitPortRegex).FindStringSubmatch(telemetry.Metrics.Address)
+		if len(explicitPortMatches) <= 1 {
+			return telemetry.Metrics.Address, defaultServicePort, nil
+		}
 
-	port, err := strconv.ParseInt(explicitPortMatches[1], 10, 32)
-	if err != nil {
-		errMsg := fmt.Sprintf("couldn't determine metrics port from configuration: %s",
-			telemetry.Metrics.Address)
-		logger.Info(errMsg, "error", err)
-		return "", 0, err
-	}
+		port, err := strconv.ParseInt(explicitPortMatches[1], 10, 32)
+		if err != nil {
+			errMsg := fmt.Sprintf("couldn't determine metrics port from configuration: %s",
+				telemetry.Metrics.Address)
+			logger.Info(errMsg, "error", err)
+			return "", 0, err
+		}
 
-	host, _, _ := strings.Cut(telemetry.Metrics.Address, explicitPortMatches[0])
-	return host, int32(port), nil
+		host, _, _ := strings.Cut(telemetry.Metrics.Address, explicitPortMatches[0])
+		return host, intToInt32Safe(int(port)), nil
+	}
+
+	port := defaultServicePort
+	host := defaultServiceHost
+	for _, reader := range telemetry.Metrics.Readers {
+		if reader.Pull != nil {
+			if reader.Pull.Exporter.Prometheus != nil {
+				if reader.Pull.Exporter.Prometheus.Host != nil && *reader.Pull.Exporter.Prometheus.Host != "" {
+					host = *reader.Pull.Exporter.Prometheus.Host
+				}
+				if reader.Pull.Exporter.Prometheus.Port != nil && *reader.Pull.Exporter.Prometheus.Port != 0 {
+					port = intToInt32Safe(int(*reader.Pull.Exporter.Prometheus.Port))
+				}
+				break
+			}
+		}
+	}
+	return host, port, nil
 }
 
 // ApplyDefaults inserts configuration defaults if it has not been set.
 func (s *Service) ApplyDefaults(logger logr.Logger) error {
-	telemetryAddr, telemetryPort, err := s.MetricsEndpoint(logger)
-	if err != nil {
-		return err
+	tel := s.GetTelemetry()
+
+	if tel == nil {
+		tel = &Telemetry{}
 	}
+	s.configureMetricReaders(logger, tel)
 
-	tm := &AnyConfig{
-		Object: map[string]interface{}{
-			"metrics": map[string]interface{}{
-				"address": fmt.Sprintf("%s:%d", telemetryAddr, telemetryPort),
-			},
-		},
+	// This field is deprecated and should not be used anymore.
+	if tel.Metrics.Address != "" {
```

Comment:
should this logic be in the upgrade package? Given we don't want to block this for older collectors being reconciled?

Reply:
On second thought, I agree. The only thing the defaulter should do is add the reader configuration if the user hasn't configured anything. If they have, we just create Service entries for it and let the upgrade take care of migrating.